### ISSUE TITLE
Calculate checksum only once per inode

### ISSUE THREAD

**Issue by kefir-** [sleuthkit/sleuthkit#397]
After trying to run tsk_loaddb on another image of a Mac backup system (Time Machine), I'm again seeing memory usage and processing times skyrocket.

As described on wikipedia at http://en.wikipedia.org/wiki/Time_Machine_%28OS_X%29#Operation
> Time Machine creates a folder on the designated Time Machine volume into which it copies the directory tree of all locally attached disk drives, except for files and directories that the user has specified to omit, including the Time Machine volume itself. Every hour thereafter, it creates a new subordinate folder and copies only files that have changed since the last backup and creates hard links to files that already exist on the backup drive.

The result of this is a LOT of hard links. Judging from the time taken and memory used, I'm guessing that tsk_loaddb calculates the checksum of every file it sees. However, it should only be necessary to calculate the checksum of each inode once, and then populate all entries from the same file system that have this inode.

Are there other optimisations that would make sense for this case? Perhaps normalizing the database to separate file name and path from the inode details like size, time stamps, permissions, etc, but I understand that this might be a big change. Checksumming each inode only once should be less invasive to fix.

For reference, here are my earlier two issues on the subject:
#373 #377

**Comment by bcarrier** [comment]
Yea, those are two ideas we've thrown around before. Separating the path versus file metadata to save space (though potentially at the expense of some queries that will require joins) and hashing only once. 

I'd say this would best be done with a flag to make the choice about checking if the file has a hash or not. For non-time machine type systems, this is not really a problem and if we always did a check, it would add a lot of database lookups and therefore take longer.

**Comment by kefir-** [comment]
I don't think it would necessarily require more database lookups, if the database structure is changed you could make tsk_files a view that looks exactly like it does today. I think sqlite should take care of any necessary joins without any problems, and still give considerable speed ups for cases like this.

For backwards compatibility, you could make tsk_files a view, and use sqlite's INSTEAD OF triggers to update the underlying tables if an application executes INSERT, UPDATE or DELETE on the view.

https://www.sqlite.org/lang_createtrigger.html#instead_of_trigger

That way, the applications using the database won't know they're using a view, and the database can be changed without breaking compatibility with old versions of the applications. 

It'd be interesting to try this out with the current tsk_loaddb actually. It should be possible to split the tables and implement tsk_files as a view with an INSTEAD OF trigger actually, and simply modify tsk_loaddb to check the database if the md5 is already set before calculating it. It'd be an interesting test case, should be fairly simple to implement, and could be tested with a Time Machine backup drive and a couple of ordinary images to see how it impacts the speed. If it seems like a good change, the tsk_file_layout table could perhaps be restructured as well to avoid duplicates.

**Comment by lfcnassif** [comment]
We are also having real problems with HFS+ images with lots of hard links (not Time Machine images). The AddImageProcess is taking more than one day and I think TSK is not calculating checksums because we are using the java bindings api and there is no option for that.

**Comment by kefir-** [comment]
I generated a test image illustrating this issue. This is a 10MB dd-image (2.1MB when compressed with gzip) containing an ext4 filesystem. I created 3 folders on it, and each folder has 1 inode with 65000 hard links (which is ext4's limit). The contents are the same in all three inodes, 1MB of null bytes.

https://drive.google.com/file/d/0B947HPmfcUmTSVE5c0I1cDV2LVk/view?usp=sharing

Processing this image with tsk_loaddb -h took 20 minutes on my machine. If it's calculating md5sum of 195000MB of data, that could be part of the explanation. Benchmarking my computer shows that 195000MB of data takes about 6 minutes to checksum when done like this:

> dd if=/dev/zero bs=1M count=195000 | time md5sum

Adding the image to Autopsy also took several minutes, and running all the ingest modules took over an hour.

(By the way, autopsy seem to show some unexpected files with names like ```/folder1/‰```, don't know why they are there.)

**Comment by lfcnassif** [comment]
Does addImageProcess, used by Autopsy, calculate checksums by default? Is possible to disable it?

**Comment by kefir-** [comment]
I don't think it does. I think the Hash Lookup ingest module calculates the checksum and then looks it up in any hash databases you have set up. Hash Lookup has an option called "Calculate MD5 even if no hash database is selected". So if you don't want to calculate checksums you can disable this ingest module.

**Comment by bcarrier** [comment]
Thanks for the link.  I'll profile that image. 

No, autopsy does not do checksum calculations at that stage.  Autopsy tries to get through the process of populating the database as fast as possible so it doesn't do hashes or store the file layout in the DB (which is why we never noticed the foreign key constraint issue that was also raised).
